File: scripts/test-peer-ping.py
#!/usr/bin/env python3
import json
import socket
import struct
import sys

host = sys.argv[1] if len(sys.argv) > 1 else "127.0.0.1"
port = int(sys.argv[2]) if len(sys.argv) > 2 else 9847
token = sys.argv[3] if len(sys.argv) > 3 else "test-token"

message = {
    "action": "ping",
    "deviceAddress": None,
    "hostName": "TestClient",
    "token": token,
}
body = json.dumps(message).encode("utf-8")
framed = struct.pack(">I", len(body)) + body

sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.settimeout(5)
sock.connect((host, port))
sock.sendall(framed)
header = sock.recv(4)
if len(header) != 4:
    print(f"FAIL: expected 4-byte header, got {len(header)}")
    sys.exit(1)
length = struct.unpack(">I", header)[0]
payload = sock.recv(length)
reply = json.loads(payload.decode("utf-8"))
print(f"OK: action={reply.get('action')} host={reply.get('hostName')}")
sock.close()

File: scripts/test-peer-ping.swift
#!/usr/bin/env swift
import Foundation
import Network

struct PeerMessage: Codable {
    enum Action: String, Codable {
        case ping, pong, status
    }
    let action: Action
    let deviceAddress: String?
    let hostName: String?
    let token: String?
}

let host = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "127.0.0.1"
let port: UInt16 = CommandLine.arguments.count > 2 ? UInt16(CommandLine.arguments[2])! : 9847
let token = CommandLine.arguments.count > 3 ? CommandLine.arguments[3] : "test-token"

let message = PeerMessage(action: .ping, deviceAddress: nil, hostName: "TestClient", token: token)
let data = try JSONEncoder().encode(message)
var framed = Data()
var length = UInt32(data.count).bigEndian
framed.append(Data(bytes: &length, count: 4))
framed.append(data)

let sem = DispatchSemaphore(value: 0)
var result = "timeout"

let connection = NWConnection(
    host: NWEndpoint.Host(host),
    port: NWEndpoint.Port(rawValue: port)!,
    using: .tcp
)

connection.stateUpdateHandler = { state in
    switch state {
    case .ready:
        connection.send(content: framed, completion: .contentProcessed { _ in
            connection.receive(minimumIncompleteLength: 4, maximumLength: 4) { header, _, _, _ in
                guard let header, header.count == 4 else {
                    result = "no response header"
                    sem.signal()
                    return
                }
                let bodyLen = header.withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
                connection.receive(minimumIncompleteLength: Int(bodyLen), maximumLength: Int(bodyLen)) { body, _, _, _ in
                    defer { sem.signal() }
                    guard let body, let reply = try? JSONDecoder().decode(PeerMessage.self, from: body) else {
                        result = "invalid response body"
                        return
                    }
                    result = "action=\(reply.action.rawValue) host=\(reply.hostName ?? "nil")"
                }
            }
        })
    case .failed(let error):
        result = "failed: \(error.localizedDescription)"
        sem.signal()
    default:
        break
    }
}

connection.start(queue: .global())
_ = sem.wait(timeout: .now() + 5)
connection.cancel()
print(result)